### PR TITLE
Enable existing service to be as `TransientService`

### PR DESCRIPTION
**Pull request by ghkim3221**
### Motivation

 - Provide a way to make existing services as `TransientService`
   - In some cases, users want to make a service not to collect metrics like a `GrpcService` which implements [GRPC Health Checking Protocol](https://github.com/grpc/grpc/blob/master/doc/health-checking.md)

### Modifications

 - Add `newDecorator` to `TransientHttpService` and `TransientRpcService`
 - Add `WrappingTransientHttpService` and `WrappingTransientRpcService` internally

### Result

 - Users can make existing service to be as `TransientService` by decorating with the `TransientHttpService` or `TransientRpcService` as follows:

   ```java
   builder.service("/docs", new DocService().decorate(
       TransientHttpService.newDecorator(TransientServiceOption.WITH_ACCESS_LOGGING)));
   ```

### CODE REVIEW

**Comment by ikhoon** [comment]
How about hiding `WrappingTransientHttpService` and `WrappingTransientHttpService` from public API and adding factory methods to `TrasientService` and `TrasientRpcService`?

For example:
```java
public interface TrasientService {
    static Function<? super HttpClient, SimpleDecoratingHttpService> newDecorator(
            TransientServiceOption... transientServiceOptions) {
        ...
    }
}
```

**Comment by ghkim3221** [comment]
Thanks! nice idea.
In fact, I also added factory methods in `TransientHttpService` and `TransientRpcService`. But I realized that they are closed for extension if there are more properties in `TransientService`. So, I changed to public API with builder.
I'm not sure my thought is right. Please consider my intention and give me advice again. I also agree to your suggestion as I have done.

**Comment by ikhoon** [comment]
> But I realized that they are closed for extension if there are more properties in TransientService.

Does that mean the factory method `newDecorator(TransientServiceOption... transientServiceOptions)` is not extendable? I think we add a builder method when we really need more options.

**Comment by ghkim3221** [comment]
Sorry for late. You right, I thought too far into the future.
Thanks @ikhoon and @trustin 👍, PTAL.

File: core/src/main/java/com/linecorp/armeria/server/TransientService.java [hunk 1]
```diff
@@ -31,6 +36,17 @@
 @FunctionalInterface
 public interface TransientService<I extends Request, O extends Response> extends Service<I, O> {
 
+    /**
+     * Returns a new {@link HttpService} decorator which makes the specified {@link HttpService} as
+     * {@link TransientService}.
+     */
+    static Function<? super HttpService, SimpleDecoratingHttpService> newDecorator(
```
Comment (minwoox) [inline core/src/main/java/com/linecorp/armeria/server/TransientService.java:43]: Should we move this to `TransientHttpService`?

File: core/src/main/java/com/linecorp/armeria/server/WrappingTransientHttpService.java [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+
+package com.linecorp.armeria.server;
+
+import java.util.Set;
+
+import com.linecorp.armeria.common.HttpRequest;
+import com.linecorp.armeria.common.HttpResponse;
+
+/**
+ * Decorated a {@link HttpService} to be treated as {@link TransientService} without inheritance.
```
Comment (minwoox) [inline core/src/main/java/com/linecorp/armeria/server/WrappingTransientHttpService.java:25]: nit: Decorates an?

File: core/src/main/java/com/linecorp/armeria/server/WrappingTransientRpcService.java [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+
+package com.linecorp.armeria.server;
+
+import java.util.Set;
+
+import com.linecorp.armeria.common.RpcRequest;
+import com.linecorp.armeria.common.RpcResponse;
+
+/**
+ * Decorated a {@link RpcService} to be treated as {@link TransientService} without inheritance.
```
Comment (minwoox) [inline core/src/main/java/com/linecorp/armeria/server/WrappingTransientRpcService.java:25]: ditto

File: core/src/main/java/com/linecorp/armeria/server/TransientService.java (hunk 1 above)
Comment (ikhoon) [reply core/src/main/java/com/linecorp/armeria/server/TransientService.java:43]: Ah... you are right. I thought this is an `HttpService` like `LoggingService`. 😅

File: core/src/main/java/com/linecorp/armeria/server/TransientService.java [hunk 4]
```diff
@@ -31,6 +36,17 @@
...
+            TransientServiceOption... transientServiceOptions) {
+        requireNonNull(transientServiceOptions, "transientServiceOptions");
+        return delegate -> new WrappingTransientHttpService(delegate,
+                                                            ImmutableSet.copyOf(transientServiceOptions));
```
Comment (ikhoon) [inline core/src/main/java/com/linecorp/armeria/server/TransientService.java:47]: - Could use `EnumSet` such as `Sets.immutableEnumSet()`?
- Add another factory method for `Iterable<TransientServiceOption> transientServiceOptions`?

File: core/src/test/java/com/linecorp/armeria/server/WrappingTransientRpcServiceTest.java [hunk 5]
```diff
@@ -0,0 +1,45 @@
...
+
+import java.util.Set;
+
+import org.junit.jupiter.api.Test;
+
+import com.linecorp.armeria.common.RpcResponse;
+
+class WrappingTransientRpcServiceTest {
+
+    static final RpcService FOO = (ctx, req) -> RpcResponse.of("foo");
```
Comment (ikhoon) [inline core/src/test/java/com/linecorp/armeria/server/WrappingTransientRpcServiceTest.java:29]: nit: A `Service` is not constant and `FOO` is not good for a service name. Just use `fooService`?

File: core/src/main/java/com/linecorp/armeria/server/TransientRpcService.java [hunk 6]
```diff
@@ -24,4 +30,15 @@
...
+    /**
+     * Returns a new {@link RpcService} decorator which makes the specified {@link RpcService} as
+     * {@link TransientService}.
+     */
+    static Function<? super RpcService, SimpleDecoratingRpcService> newDecorator(
+            TransientServiceOption... transientServiceOptions) {
+        requireNonNull(transientServiceOptions, "transientServiceOptions");
+        return delegate -> new WrappingTransientRpcService(delegate,
+                                                           ImmutableSet.copyOf(transientServiceOptions));
+    }
```
Comment (ikhoon) [inline core/src/main/java/com/linecorp/armeria/server/TransientRpcService.java:43]: Ditto
- Use `EnumSet`?
- Add a new method that takes `Iterable`.

**Comment by ghkim3221** [comment]
Thanks, @ikhoon and @minwoox! PTAL 😄

**Review by ikhoon — APPROVED** [review]
Thanks! @ghkim3221 🎄🎅🏻

**Review by trustin — APPROVED** [review]
Thanks, @ghkim3221 !

**Review by minwoox — APPROVED** [review]
Thanks, @ghkim3221!

**Comment by trustin** [comment]
@ghkim3221 Would you mind updating the PR description so it matches what we have now?

**Comment by ghkim3221** [comment]
Sure

**Comment by trustin** [comment]
Thanks, @ghkim3221. Merged :tada:

**Pull request merged**